Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/MarkerDetection/QRCode/QRCodeMarkerDetector.cs

```diff
@@ -0,0 +1,257 @@
+        {
+            if (_tracking &&
+                _processMarkers)
+            {
+                ProcessMarkerUpdates();
+            }
+        }
+
+        private void QRCodeAdded(object sender, QRCodeEventArgs<QRCodesTrackerPlugin.QRCode> e)
+        {
+            int markerId;
+            lock (_cachedMarkers)
+            {
+                if (!_cachedMarkers.ContainsKey(e.Data.Id) &&
+                    TryGetMarkerId(e.Data.Code, out markerId))
+                {
+                    _cachedMarkers.Add(e.Data.Id, new QRCodeInfo(e.Data.Id, markerId, e.Data.PhysicalSizeMeters));
+                    _processMarkers = true;
+                }
+            }
+        }
+
+        private void QRCodeUpdated(object sender, QRCodeEventArgs<QRCodesTrackerPlugin.QRCode> e)
+        {
+            int markerId;
+            lock (_cachedMarkers)
+            {
+                if (!_cachedMarkers.ContainsKey(e.Data.Id) &&
+                TryGetMarkerId(e.Data.Code, out markerId))
+                {
+                    _cachedMarkers.Add(e.Data.Id, new QRCodeInfo(e.Data.Id, markerId, e.Data.PhysicalSizeMeters));
+                    _processMarkers = true;
+                }
+            }
+        }
+
+        private void QRCodeRemoved(object sender, QRCodeEventArgs<QRCodesTrackerPlugin.QRCode> e)
+        {
+            lock(_cachedMarkers)
+            {
+                if (_cachedMarkers.ContainsKey(e.Data.Id))
+                {
+                    _cachedMarkers.Remove(e.Data.Id);
+                    _processMarkers = true;
+                    if (_markerCoordinateSystems.ContainsKey(e.Data.Id))
+                    {
+                        _markerCoordinateSystems.Remove(e.Data.Id);
+                    }
+                }
+            }
+        }
+
+        private void ProcessMarkerUpdates()
+        {
+            lock (_cachedMarkers)
+            {
+                foreach (var markerPair in _cachedMarkers)
+                {
+                    if (!_markerCoordinateSystems.ContainsKey(markerPair.Key))
+                    {
+                        var coordinateSystem = SpatialGraphInteropPreview.CreateCoordinateSystemForNode(markerPair.Key);
+                        if (coordinateSystem != null)
+                        {
+                            _markerCoordinateSystems.Add(markerPair.Key, coordinateSystem);
+                        }
+                    }
+                }
+            }
+
+            bool locatedAllMarkers = true;
+            var markerDictionary = new Dictionary<int, Marker>();
+            foreach (var coordinatePair in _markerCoordinateSystems)
+            {
+                int markerId = -1;
+                lock (_cachedMarkers)
+                {
+                    if (_cachedMarkers.ContainsKey(coordinatePair.Key))
+                    {
+                        markerId = _cachedMarkers[coordinatePair.Key].MarkerId;
+                    }
+                }
```

Comment:
Couple of points here:
- If you are locking, you should lock around both dictionaries/managers if they are meant to be in sync
- I would use TryGetValue, and it it returns false do the Debug.Log, set locatedAllMarkers to false and then "continue;"

Reply:
moved to a single content lock. updated code cleanliness to reflect less locking